### PR TITLE
Allow drag and drop to create Rows and Galleries

### PR DESCRIPTION
**Pull request by tellthemachines**
## What?

Implements #13202

Enhances drag and drop so that, when hovering over or around blocks in a vertical layout, dropzones appear around the sides to allow for grouping blocks horizontally into Rows. 

Additionally, if an Image block is dropped beside another Image block, they will transform into a Gallery. 

Is this a good addition? Feedback appreciated!

## Testing Instructions

1. Open a post and multiple blocks;
2. Try dragging them around each other to see the lateral dropzones appear;
3. Try dropping a block into a lateral dropzone and check that a Row is created with both blocks;
4. Drag an image from outside into the editor and check that it can also be formed into  a Row with existing blocks;
5. Drag an Image block beside another Image block and check that they become a Gallery block;
6. Try to break this in imaginative ways.

Todo: 
* make images dragged in from outside onto Image blocks become Galleries too (currently they just become Rows)
* address test failures

https://github.com/WordPress/gutenberg/assets/8096000/bcc04337-5e0e-437c-b4af-d3cab022e7b2

### CODE REVIEW

**Comment by alexstine** [comment]
@tellthemachines Please let Accessibility Team know when this is ready for early feedback.

From personal memory, this kind of sounds like an accessibility nightmare. When I used to magnify the screen and have to drag and drop, overlapping used to be a common problem. I can just see this turning into a situation where users perform actions they do not expect. Especially on zoomed in views.

With that said, this is still draft but wanted to leave this note while it's still early.

Thanks.

**Comment by tellthemachines** [comment]
That's a great point @alexstine, thanks for the reminder! I'll have a play around with this in a magnified screen and see how it works.
This is very much an experiment and one of the goals here is to uncover any potential usability issues or conflicts with existing functionality. I'm happy to get feedback at any point, but especially once I work out how to drag and drop blocks onto other blocks. I'll add the a11y feedback label to it once it's working roughly as expected!

**Comment by andrewserong** [comment]
Very cool exploration! I wonder how far potentially unexpected behaviour could be mitigated by having a live preview of what's going to happen prior to a user letting go of the mouse button? So if a user goes to drag over a block and it'll turn it into a Row block, it shows what that'll look like before letting go. Another thing that could potentially help there is if only a particular part of the block you're dragging over activated the behaviour where it switches to a row — e.g. if you're hovering over the right edge of a paragraph block then it'll switch to Row and add to the right, but not if you're just dragging over the middle of the block.

That's just a couple of quick thoughts / first impressions, though. Great idea playing around with this one!

**Comment by tellthemachines** [comment]
Added "Needs accessibility feedback" label as per @alexstine 's suggestion; this is an early experiment but would be great to know if there may be any major issues with the approach. 

Preliminary testing on zoomed in view shows functionality is not available on over 200% due to the block toolbar not displaying the drag handle when it's attached to the top of the canvas (which is its default mode of display on small screen/high zoom level). I'm not sure if the same will happen when using a magnifying app; would love some pointers regarding common apps to test with! I have both Mac and Windows so could test on either.

**Comment by mtias** [comment]
Nice! I think this could make cases like "add two images side by side" a lot more intuitive to accomplish. I think how we communicate the result needs to be tweaked, though—not with an overlay on top of the entire block. I think we need some design options there.

**Comment by tellthemachines** [comment]
Note that the latest commit changes the inserter type to inbetween BUT I haven't tweaked its placement yet so it's popping up in all sorts of weird places 😅 

Todo: try showing inserter to the left or right of the block being dragged over depending on current hover position.

**Comment by jasmussen** [comment]
Catching up let me see if I can address the "Needs Design" label. Here's a mockup:

Drag and Drop

Shown in this mockup is a header layout with a site title. In the next slide, a photo is dragged on top of this site title block, as that happens, the site title gets highlighted with a 10% blue background, and a vertical blue line appears in front of the site title.

The idea here is that the site title is in a Row container, so dragging any block into it will create a horizontal layout. Dragging on the left side of the site title will show this blue drop-indicator to the left. If you drag the image to the right side of the title, it will show the drop indicator to the right.

In other words, this is a drag and drop equivalent of the "Add before" or "Add after" options, with a visual indicator created by the block you're dragging on top of. Hope that helps!

**Comment by tellthemachines** [comment]
I've updated this draft to use the drop indicator roughly as shown in @jasmussen 's designs. I say roughly because there is a slight obstacle to this behaviour when trying to drop over blocks that aren't already in a horizontal layout: most blocks are full-width by default, so when the inserter appears on the right hand side, it's to the right of the whole element box, and if the block content doesn't stretch to full width (as it won't with small images or some headings) then the indicator can be pretty far away from the block.

I'm not sure how to deal with this as afaik there's no way of retrieving the width of just the content without changing the block style to inline.

Here's a video showing current behaviour:

https://github.com/WordPress/gutenberg/assets/8096000/d826f828-7cd1-4d26-a6b3-8b152556497f

(I haven't added the 10% blue background highlight yet)

**Comment by jasmussen** [comment]
Nice! Thank you for working on this. I'm almost certainly missing nuance, so I'm going to list out some thoughts, then you can correct me or point out the misconceptions:

* When the container block is a horizontal flex container (such as Row), show the drop-zones left and right. 
* When the container block is a vertical flex container (such as Stack), show the drop-zones above and below
* Similarly, when the container is a flow container (such as Group, or a Column), show the drop-zones above and below

Following that train of thought, in this case:

> so when the inserter appears on the right hand side, it's to the right of the whole element box, and if the block content doesn't stretch to full width (as it won't with small images or some headings) then the indicator can be pretty far away from the block

In this case, the block is either "filling" the remaining part of a horizontal flex box, and the blue indicator all the way to the right would be correct, since that's where the block will end up.

There's another behavior too. In this GIF are a bunch of paragraphs, and an image. When I drag the image on top of a preceding paragraph, dropzones left and right are surfaced, and when I drop on a zone left, the paragraph is turned into a row, with the image on the left:

drag on top of text

In principle, that's a cool layout tool, but it's not actually what I would expect. I wouldn't expect any containers to be created as part of this effort, so if I drag an image onto a paragraph in a default flow-based post container, I would expect the dropzones to be highlighted _before_ and _after_ the paragraph, like this:

Drag and Drop

_I realize I should've shared this instinct earlier on, apologies._

One way I was thinking about this was that we actually output these drop-zones in the DOM, before and after the block hovered. The benefit would be, presumably, that those drop-zones in the DOM would both _make the space available_, but it would also intrinsically place itself where it needs to be as it would be subject to the container flex or flow rules. The main problem here is the layout shift this will cause, you might enter into a loop where the block you're hovering would layout-shift and you'd no longer be hovering, but I wonder if we can be smart there, and remember which block is hovered so long as the dropzones exist, and not blur it until you're also not hovering any of the dropzones?

**Comment by tellthemachines** [comment]
> When I drag the image on top of a preceding paragraph, dropzones left and right are surfaced, and when I drop on a zone left, the paragraph is turned into a row, with the image on the left:
> In principle, that's a cool layout tool, but it's not actually what I would expect. I wouldn't expect any containers to be created as part of this effort

Apologies if I didn't make the intent fully clear 😅 this exploration is based on the ideas in #13202 which aim to create appropriate layout containers when users try to drag blocks next to other blocks. Essentially we're trying to guess when the desired outcome is to have the two blocks side by side, and wrap them in a Row. Do you think the visual indicators from your designs could be made to work for such a case?

Regarding making the before and after dropzones visible while hovering over a block, @andrewserong has an experiment over in #56466.

**Comment by jasmussen** [comment]
Oh, understood. 

As a prototype of that this PR certainly gets close. But I will also say in testing, that this didn't feel as useful in practice as it might sound. I think there's still room for the add-before suggestion as outlined. Perhaps they can even coexist — hover top or bottom halfs, it's before/after (like Add before), hover on the left or right halves, it's _convert to row and put before or after, unless it's already in a Row in which case just put before or after_. Make sense?

**Comment by tellthemachines** [comment]
Ok peeps I think the experience here is now reasonably nice 😄 so I've marked this ready for review.

File: packages/block-editor/src/components/use-on-block-drop/index.js [hunk 1]
```diff
@@ -222,12 +243,54 @@ export default function useOnBlockDrop(
...
+						block.innerBlocks
+					);
+				} );
+
+				const areAllImages = blocks.every( ( block ) => {
+					return block.name === 'core/image';
+				} );
+
+				const wrappedBlocks = createBlock(
+					areAllImages ? 'core/gallery' : 'core/group',
```
Comment (andrewserong) [inline packages/block-editor/src/components/use-on-block-drop/index.js]: Since we're in the block-editor package, do we need to check that the `core/gallery` and `core/group` blocks are available? Just thinking of other 3rd party editors that might wind up using the `block-editor` package.

For the Group block, there's the `isGroupable` selectors, which might be what we're after: https://github.com/WordPress/gutenberg/blob/a8b4cc216c68fa1c42d25e0911d3cdbe8e1e619e/docs/reference-guides/data/data-core-block-editor.md#L1080-L1082

Then, in place of the hard-coded string for `core/group`, we could use `getGroupingBlockName`, similar to how the block actions handles it? https://github.com/WordPress/gutenberg/blob/1710facd57ad89595c6b0dedc5f30cd0f4e694d3/packages/block-editor/src/components/block-actions/index.js#L111

For the Gallery block, I suppose we could do something like `!! getBlockType( 'core/gallery' )` to check that it's available?

**Review by andrewserong — COMMENTED** [review]
Very cool! This is generally testing pretty well for me. I ran into a couple of issues:

* When rearranging paragraphs within the editor canvas, it feels like the left and right zones are now very prominent and I found it easy to for them to get in the way of vertical re-ordering. Would it be worth applying the left/right behaviour only within a particular threshold, a bit like we did for dragging between adjacent blocks?
* When I drag to the very right of a full-width block in the site editor, there's a point at which the width of the editor canvas flickers with an additional scrollbar. In case it doesn't show for you during testing, I have my system set to show scrollbars always:

image

Here's a screengrab of looking at re-ordering a paragraph, and at the end the flickering at the edge of the viewport:

https://github.com/WordPress/gutenberg/assets/14988353/c139661a-4d2b-4ab9-a34e-ad32b42e8f16

Another tricky one is that it's possible to drag from a Row block to the edge of a Row block and then wind up with an additional level of nesting unexpectedly. Here's how I could wind up in that state:

https://github.com/WordPress/gutenberg/assets/14988353/d4511608-dac7-4217-8156-b926f1d68f09

For that last one, do we need to do more checking to see if it looks like something has already been grouped? 🤔

File: packages/block-editor/src/components/use-on-block-drop/index.js (hunk 1 above)
Comment (tellthemachines) [reply packages/block-editor/src/components/use-on-block-drop/index.js]: Oh yes definitely, that's a great point!

**Comment by tellthemachines** [comment]
>  Would it be worth applying the left/right behaviour only within a particular threshold, a bit like we did for dragging between adjacent blocks?

I kind of did this but it's probably too big a threshold 😅 given it's 1/4 of the block width.

Do you reckon `THRESHOLD_DISTANCE` would be a good fit for this? Or should it be another value? This would only apply when hovering over the inside of the block; from the outside I think it should work regardless of the distance.

**Comment by andrewserong** [comment]
> Do you reckon THRESHOLD_DISTANCE would be a good fit for this?

I think it's probably a good value to try. It's hopefully big enough that folks will find it when dragging intentionally, without accidentally dropping on it 🤞. I think things like thresholds often come down to "feel" while playing around, though, so there's probably an intuitive sweet spot somewhere 😄

**Comment by tellthemachines** [comment]
> When I drag to the very right of a full-width block in the site editor, there's a point at which the width of the editor canvas flickers with an additional scrollbar.

Ok I think know the cause of this: because your Paragraphs in that example are nested in a Group which has its edges right up against the edge of the canvas, when dragging over that area the side drop indicator for the Group goes into a state of indecision over whether to show or not. What I don't understand is why the indicator flickering should trigger a scrollbar 🤔

**Comment by tellthemachines** [comment]
Ok I think I've addressed all the feedback so far:

* scrollbar flickering fixed
* don't create duplicate Row when dropping next to an existing Row
* smaller threshold for dragging inside a block
* check that Gallery block exists and use whatever grouping block is default

I don't think we need to check that the blocks being grouped are groupable, since grouping blocks should allow all blocks? Unless they can't be dropped in which case we won't see a drop indicator now that #56843 is in place. Or is there any other scenario I'm missing? We're already doing a ton of checks here so would be good to avoid unnecessary ones 😅 

The failing tests will have to be overhauled because they were built on the assumption that dragging around the sides of a vertical block list would activate the vertical inbetween inserters, and that behaviour has now changed. Would be good to be sure this _is_ the desired behaviour before going ahead with that! I'd especially appreciate design feedback.

Would also be great to get accessibility feedback to make sure this isn't breaking anything! The only new behaviour here is the creation of a new wrapper block when blocks are dropped into certain positions, so I'd expect that to not be a huge deviation from the accessibility of the existing implementation, but I'm probably missing things here 😅

**Comment by andrewserong** [comment]
> I don't think we need to check that the blocks being grouped are groupable, since grouping blocks should allow all blocks?

We might still want an [`isGroupable` check](https://github.com/WordPress/gutenberg/blob/a8b4cc216c68fa1c42d25e0911d3cdbe8e1e619e/packages/block-editor/src/store/selectors.js#L2926) in order to ensure that we don't wrap blocks that shouldn't be allowed to be wrapped (e.g. locked blocks)? With this PR it looks possible to wrap a Group block that is locked against movement, which isn't allowed on trunk:

https://github.com/WordPress/gutenberg/assets/14988353/f567a63d-215d-41cf-8a5e-4dc017844cdc

Otherwise, the threshold is feeling nice to me in testing!

The main issue I ran into is this weird one when dragging a paragraph to the right of a Row block. It then wound up in a strange state where there's two instances of the one paragraph block, so they both appear to be selected at the same time, with duplicated block tools. I've never seen anything like it! Let me know if this is just an odd thing I've run into in my environment, and I can try to come up with some more detailed replication steps 😅

https://github.com/WordPress/gutenberg/assets/14988353/a8827347-fda9-4fe3-b2e2-eefd24a3f223

**Comment by SaxonF** [comment]
This is working really well for me, nice one @tellthemachines !

The only feedback I have is could maybe increase the threshold _ever so slightly_ **but** I think we can start with what we have here so its less intrusive and then bump it up based on feedback after release.

https://github.com/WordPress/gutenberg/assets/1072756/9ba73ab0-699c-45bd-867a-63e828f21022

**Comment by tellthemachines** [comment]
Thanks for the reviews!

> The main issue I ran into is this weird one when dragging a paragraph to the right of a Row block. It then wound up in a strange state where there's two instances of the one paragraph block, so they both appear to be selected at the same time, with duplicated block tools.

Turns out the problem was the previous logic for avoiding double-wrapped Rows. I updated to instead prevent the drop indicator from appearing beside a Row (so you should only be able to drag _into_ the Row, not beside it) and that fixed the issue for me.

Also added the groupability check and updated the unit tests to take this new behaviour into account. I think this is ready for a final check!

**Comment by andrewserong** [comment]
Nice, this is feeling a lot closer to me! The issue with nested groups and locked blocks appears to be resolved for me now 👍

I'm still getting a bit of weirdness when dragging to the very edge of a full-width block. An example is when dragging a heading block within the site editor in the TT4 blog home template where we've got a lot of quite tall Group blocks. It seems to flicker a bit with showing/hiding scrollbars for me:

https://github.com/WordPress/gutenberg/assets/14988353/bc2c6436-8b43-4ff7-b5f1-9b1bdbbfcaaf

In case it helps debug, my OS is set to always display scrollbars. I'm wondering if the vertical drop indicator line in this case is stretching off the edge of the container, and therefore creating an additional scrollbar? I wonder if we can constrain that drop indicator line so it never (technically) goes off the edge of the screen 🤔 

To distil this use case: dragging and dropping to the left or right of a long block that extends beyond the height of the viewport and might be full width.

Other than that, it looks like some of the e2e test failures might be real for `test/e2e/specs/editor/blocks/paragraph.spec.js` in Playwright 2 as the tests are looking for drop zones in particular?

**Comment by tellthemachines** [comment]
Thanks for re-reviewing! I've fixed the e2e and added a new test for this case.

> To distil this use case: dragging and dropping to the left or right of a long block that extends beyond the height of the viewport and might be full width.

That's annoying, but not specific to this branch. I can reproduce it on trunk, in the site editor, with a very long block at the edge of a Row.

<details>
<summary>Example markup:</summary>

```
<!-- wp:paragraph -->
<p>block</p>
<!-- /wp:paragraph -->

<!-- wp:group {"layout":{"type":"flex","flexWrap":"nowrap"}} -->
<div class="wp-block-group"><!-- wp:paragraph -->
<p>fha;sdfh;DS Hsdf hsck this shim.nsfglkdsjfg okajgf 'ojG 'OJDFG ;OSDHG 'osdhfg ;aodgh ;ksKDGH s;kd t</p>
<!-- /wp:paragraph -->

<!-- wp:cover {"overlayColor":"accent","isUserOverlayColor":true,"minHeight":2001,"minHeightUnit":"px"} -->
<div class="wp-block-cover" style="min-height:2001px"><span aria-hidden="true" class="wp-block-cover__background has-accent-background-color has-background-dim-100 has-background-dim"></span><div class="wp-block-cover__inner-container"><!-- wp:paragraph -->
<p>hsgadgf ad</p>
<!-- /wp:paragraph -->

<!-- wp:paragraph -->
<p>g adf ad</p>
<!-- /wp:paragraph --></div></div>
<!-- /wp:cover --></div>
<!-- /wp:group -->
```

</details>

I think it might be a good candidate for exploring a fix for separately, especially given it's a bit of an edge case.

**Review by andrewserong — APPROVED** [review]
> That's annoying, but not specific to this branch. I can reproduce it on trunk, in the site editor, with a very long block at the edge of a Row.

Oh, great catch! Yes, I can reproduce that on trunk now, but bizarrely only in the site editor and not in the post editor for some reason 🤔

In any case, it seems this PR just makes the issue more visible since we're using the vertical drop indicator more frequently, so I agree, that sounds like a good bug for us to look into fixing separately.

This looks good to try to me!

Nice work getting this one over the line ✨

**Comment by tellthemachines** [comment]
> I can reproduce that on trunk now, but bizarrely only in the site editor and not in the post editor

Same! I wonder if there's some difference in how the iframe is setup between editors 🤔 

Anyway, thanks for the thorough testing!

**Comment by Mamaduka** [comment]
Excellent work; I love the feature!

**Comment by Mamaduka** [comment]
@tellthemachines, I just noticed that one of the tests became flaky after this got merged - #58161.

My attempt to fix it - #58208.

**Comment by t-hamano** [comment]
@tellthemachines 

It seems that this PR causes an error to occur when dragging and dropping multiple local images into the editor.

See #58653
